torch/csrc/jit/argument_spec.h

```diff
@@ -0,0 +1,180 @@
+#include <Python.h>
+#include <iostream>
+#include <vector>
+#include "torch/csrc/autograd/variable.h"
+#include "torch/csrc/utils/hash.h"
+
+namespace torch { namespace jit {
+
+// GraphExecutor creates specializations of Graphs for different dimensionalitities
+// and types of inputs.
+
+// ArgumentSpec represents one particular specialization.
+// It is designed so that it can be created, hashed, and compared quickly
+// since it is used along the hot-path of the JIT to check if the code
+// we have created is valid for the given inputs.
+
+// TensorInfoPOD is only used internally in ArgumentSpec
+// API users should use TensorInfo
+struct TensorInfoPOD {
+  // total size is 64-bit
+  unsigned int type : 8;
+  unsigned int defined : 1;
+  unsigned int requires_grad : 1;
+  int device : 22;
```

Comment:
Is this really well defined to store a *signed* int in 22 bits?

Reply:
Aaaand there you go:
> Whether char, short, int, long, and long long bit fields that aren't explicitly signed or unsigned are signed or unsigned. For example, `int b:3`; may have the range of values `0..7` or `-4..3`. (until C++14)

Reply:
Does that mean that if I say 'signed int', then it is ok?

Reply:
Seems like yes: https://stackoverflow.com/questions/149869/does-ansi-c-support-signed-unsigned-bit-fields